### ISSUE TITLE
opendkim says key data is not secure

### ISSUE THREAD

**Issue by binarykitchen** [mail-in-a-box/mailinabox#1552]
Seeing this in my syslog

```
Apr 11 14:18:46 xxx opendkim[5426]: xxx.com: key data is not secure: /home/user-data is writeable and owned by uid 1002 which is not the executing uid (118) or the superuser
```

Any clues? I didn't change anything from default mailinabox configuration.

**Comment by alento-group** [comment]
This is a result of MiaB being configured with a custom path to the specific key file.  That file location is not owned by the system user 'opendkim' or by the root user so the error is shown as opendkim expects the file to be elsewhere.

**Comment by alento-group** [comment]
Further more in depth explanation taken from: 
[DKIM is not secure issue](https://lujihood.com/dkim-authentication-problem-on-mailplus/)



The safety check you’re hitting attempts to determine if any user other than root and the user running the opendkim binary could conceivably alter the key file you’re trying to use. Every directory from the root down is checked, as is the key file itself. If any of them could be written by a user other than those two, the error appears. 

Specifically, the error is reporting that it found a directory in group 6 either containing your key or someplace above it that is group-writable, and also found that there’s some other account (“pop” in this case) that is in group 6. That could be determined by looking for other accounts with primary group 6 in the password file (using getpwent()), or via a query to group 6 in the groups file (using getgrgid()).

So, in short, opendkim is alarming because if someone can hack the “pop” user, they can also alter your DKIM keys. If changing the permissions for the offending directory isn’t possible, you can suppress the check by setting “RequireSafeKeys” to False in your configuration file.

    -MSK

**Comment by alento-group** [comment]
> So, in short, opendkim is alarming because if someone can hack the “pop” user, they can also alter your DKIM keys. If changing the permissions for the offending directory isn’t possible, you can suppress the check by setting “RequireSafeKeys” to False in your configuration file.

@JoshData this suggests a proposed fix for this issue.

**Comment by JoshData** [comment]
We already are doing that.

**Comment by binarykitchen** [comment]
@JoshData already doing that? where? A separate issue or in a future version?

**Comment by JoshData** [comment]
https://github.com/mail-in-a-box/mailinabox/search?utf8=%E2%9C%93&q=RequireSafeKeys&type=

**Comment by binarykitchen** [comment]
If we are already doing that, then why am I still seeing this warning? Sorry if the question is silly but I'm trying to understand what's going on.

**Comment by lorrin** [comment]
I know this is an old issue, but an explanation was never posted. The [opendkim man page](http://www.opendkim.org/opendkim.conf.5.html) states:

> RequireSafeKeys (boolean)
> When reading a key file, a message will be logged if the key file has the read or write bit set other than for the owner or for a group that the executing process is in. With this feature set to "true", the filter will further consider this an error and refuse to make use of the file’s contents. The default is "true".

In other words, the setting determines whether it is considered a fatal error or not, but the warning is logged either way.

**Comment by monarc83** [comment]
I have also just stumbled across these log messages. Is this still a security risk or has it already been fixed and the log message can be ignored?

**Comment by JoshData** [comment]
It was never a real security risk.
